glpi-project/glpi
# Cypress POC

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Suggested alternative to Symfony/Panther (#14999).
This PR currently contains no meaningful/real tests. It **does** however show that it is capable of some things that were difficult/impossible with Panther like typing within a TinyMCE editor or interacting with flatpickr.
Also, to say docker/github actions are something I'm not good with would be an understatement. So, the base URL is hard-coded for my development instance currently in the config rather than starting GLPI in a container using a test database. The tests done currently don't make any changes to the data except changing your current ticket search. **I will need help getting it working in a containerized environment like the other tests.**

- Flatpickr was handled with a custom `selectDate` command. This command can select the date by either interacting with flatpickr itself (default) or simply typing the date depending on the "interactive" option.
- TinyMCE was handled by hooking into Cypress' `type` command and detecting if it is a textarea with a tinymce editor. I added examples of using the `setContent` function on the editor as well as interactively typing within the editor.
- Select2 works completely without issue as long as you use the "force" option with the `select` command. The "force" option tells Cypress to ignore the visibility checks it does. Another option could be hooking into the command like I did for `type` to detect select2 and handle it without requiring the "force" option.

Finally, I added a `test:e2e` npm command which runs Cypress in interactive mode (`cypress open`). The tests can be run in headless mode with `cypress run` instead. When the Cypress window opens, you will need to select the "E2E" option. When the tests are run in interactive/headed mode, you can navigate through each test/step and see a screenshot of what the screen looked like at the time.

`cypress install`
`npm run test:e2e`
- Select E2E Testing
- Pick the browser you want to run the tests in. Depends on what you have installed. My options were Chrome, Electron and Firefox. There is a way to add experimental Webkit (Safari) support later if desired through `playwright-webkit`.
- Select the "poc.cy.js" spec
- Wait and watch the tests run. You can see the actual browser window as Cypress runs the tests and interacts with it.
- In the second panel on the left you can see a list of the tests that were run. You can click them to expand them to see the individual steps. Hovering over the steps will show you a preview of what the screen looked like (not perfect).
